src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java

```diff
@@ -0,0 +1,55 @@
+package org.ibp.api.java.impl.middleware.study;
+
+import org.generationcp.middleware.service.api.PedigreeService;
+import org.generationcp.middleware.service.api.study.germplasm.source.GermplasmStudySourceDto;
+import org.generationcp.middleware.service.api.study.germplasm.source.GermplasmStudySourceRequest;
+import org.generationcp.middleware.util.CrossExpansionProperties;
+import org.ibp.api.java.impl.middleware.study.validator.StudyValidator;
+import org.ibp.api.java.study.GermplasmStudySourceService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.annotation.Resource;
+import java.util.List;
+
+@Service
+@Transactional
+public class GermplasmStudySourceServiceImpl implements GermplasmStudySourceService {
+
+	@Resource
+	private StudyValidator studyValidator;
+
+	@Resource
+	private PedigreeService pedigreeService;
+
+	@Resource
+	private CrossExpansionProperties crossExpansionProperties;
+
+	@Resource
+	private org.generationcp.middleware.service.api.study.germplasm.source.GermplasmStudySourceService
+		germplasmStudySourceMiddlewareService;
+
+	@Override
+	public List<GermplasmStudySourceDto> getGermplasmStudySourceList(final GermplasmStudySourceRequest germplasmStudySourceRequest) {
+
+		this.studyValidator.validate(germplasmStudySourceRequest.getStudyId(), false);
+
+		final List<GermplasmStudySourceDto> germplasmStudySourceDtoList = this.germplasmStudySourceMiddlewareService
+			.getGermplasmStudySourceList(germplasmStudySourceRequest);
+
+		for (final GermplasmStudySourceDto germplasmStudySourceDto : germplasmStudySourceDtoList) {
+			germplasmStudySourceDto.setCross(this.pedigreeService.getCrossExpansion(germplasmStudySourceDto.getGid(), this.crossExpansionProperties));
```

Comment:
@darla-leafnode will this perform well? we had some issues calling the cross expansion before.

Reply:
In our testing, this performs fast as we only process the records on the current page (100 max). But we can further improve this by using this.pedigreeService.getCrossExpansions to minimize the expense of calling crossExpansionProperties.getCropGenerationLevel multiple times.

Reply:
Done